packages/core/src/server/service.ts

```diff
@@ -1,197 +1,181 @@
-        };
-        return res.status(503).json(result);
+          errors: [new GraphQLError("Historical indexing in not complete")],
+        });
       }
 
-      switch (req.method) {
-        case "POST":
-          return this.graphqlMiddleware(req, res, next);
-        case "GET": {
-          return res
-            .status(200)
-            .setHeader("Content-Type", "text/html")
-            .send(graphiQLHtml);
-        }
-        case "HEAD":
-          return res.status(200).send();
-        default:
-          return next();
+      if (c.req.method === "POST") {
+        const getLoader = buildLoaderCache({ store: indexingStore });
+
+        c.set("store", indexingStore);
+        c.set("getLoader", getLoader);
+
+        return graphqlServer({
+          schema: graphqlSchema,
+        })(c);
+      }
+      return next();
+    })
+    .get("/graphql", (c) => {
+      return c.html(graphiQLHtml);
+    })
+    .use("/", async (c, next) => {
+      if (c.req.method === "POST") {
+        const getLoader = buildLoaderCache({ store: indexingStore });
+
+        c.set("store", indexingStore);
+        c.set("getLoader", getLoader);
+
+        return graphqlServer({
+          schema: graphqlSchema,
+        })(c);
+      }
+      return next();
+    })
+    .get("/", (c) => {
+      return c.html(graphiQLHtml);
+    });
+
+  const createServerWithNextAvailablePort: typeof http.createServer = (
+    ...args: any
+  ) => {
+    const httpServer = http.createServer(...args);
+
+    const errorHandler = (error: Error & { code: string }) => {
+      if (error.code === "EADDRINUSE") {
+        common.logger.warn({
+          service: "server",
+          msg: `Port ${port} was in use, trying port ${port + 1}`,
+        });
+        port += 1;
+        setTimeout(() => {
+          httpServer.close();
+          httpServer.listen(port, common.options.hostname);
+        }, 5);
       }
     };
-  }
+
+    const listenerHandler = () => {
+      common.metrics.ponder_server_port.set(port);
+      common.logger.info({
+        service: "server",
+        msg: `Started listening on port ${port}`,
+      });
+      httpServer.off("error", errorHandler);
+    };
+
+    httpServer.on("error", errorHandler);
+    httpServer.on("listening", listenerHandler);
+
+    return httpServer;
+  };
+
+  const httpServer = await new Promise<http.Server>((resolve, reject) => {
+    const timeout = setTimeout(() => {
+      reject(new Error("HTTP server failed to start within 5 seconds."));
+    }, 5_000);
+
+    const httpServer = serve(
+      {
+        fetch: hono.fetch,
+        createServer: createServerWithNextAvailablePort,
+        port,
+        // Note that common.options.hostname can be undefined if the user did not specify one.
+        // In this case, Node.js uses `::` if IPv6 is available and `0.0.0.0` otherwise.
+        // https://nodejs.org/api/net.html#serverlistenport-host-backlog-callback
+        hostname: common.options.hostname,
```

Comment:
It's a bit suspicious to me that we're passing in hostname here as well as above, but it should work fine.